### ISSUE TITLE
Own material video, discounted effects, distortion

### ISSUE THREAD

**Issue by SubePython** [cfzd/Ultra-Fast-Lane-Detection#69]
Brother, may I ask:
 1. I use my own lane line material video. After using the demo visualization, the visualization effect is very poor and the dislocation is serious. Is the video resolution different from the resolution of the culane dataset, or do you need to change something? Please give me  Some suggestions, thanks
大佬，请问一下：
1. 我使用自己的车道线素材视频，使用的demo可视化后，可视化效果特别差，错位严重，是视频分辨率跟culane数据集分辨率不同导致的吗，或者需要更改些什么吗，请给我一些建议，谢谢
![Screenshot_20200812-210832](https://user-images.githubusercontent.com/39877932/90018793-0d2c1800-dce0-11ea-9077-27c89360ca96.jpg)
![Screenshot_20200812-210817](https://user-images.githubusercontent.com/39877932/90018804-10270880-dce0-11ea-87ad-c2867f1dd2e7.jpg)
![Screenshot_20200812-210756](https://user-images.githubusercontent.com/39877932/90018806-10bf9f00-dce0-11ea-8fc4-a37cff100db9.jpg)

**Comment by cfzd** [comment]
@SubePython 
It should be the size misalignment of sizes of your input video and the defalut size in `demo.py`. The fastest way of solving  this problem is to resize your video to `1640x590` or `1280x720`. For `1640x590`, you can use the config of CULane dataset. For `1280x720`, you can use the config of Tusimple.

**Comment by SubePython** [comment]
@cfzd 
我尝试改变我视频的大小与culane相同，效果依然不理想，后尝试使用tusimple的大小，简直看不了，我的视频大小为1280x720
I tried to change the size of my video to be the same as that of culane, but the effect was still not ideal. After that, I tried to use the size of tusimple. I could hardly see it. My video size was 1280 x 720
![微信截图_20200816201916](https://user-images.githubusercontent.com/39877932/90334164-1bd73f80-dffe-11ea-8528-6206c1c6d34d.png)

**Comment by cfzd** [comment]
@SubePython 
Can you show me your test code? Especially the image reading part, image transformation part, and `cv2.circle` call part.

**Comment by SubePython** [comment]
@cfzd 
感谢您的耐心解答：
1. 代码采用的是issue#22，大佬的代码
2. 测试视频：
 链接：https://pan.baidu.com/s/1QvoX7iRSwMw8YBZtzmKSlA 
提取码：0cnb 
3. 还请指正我错误的地方

Thank you for your patience
1. The code uses issue ා 22, the code of big man
2. Test video:
Link: https://pan.baidu.com/s/1QvoX7iRSwMw8YBZtzmKSlA
Extraction code: 0cnb
Please point out my mistakes

# -*- coding: utf-8 -*-

import torch, os, cv2
from model.model import parsingNet
from utils.common import merge_config
from utils.dist_utils import dist_print
import torch
from PIL import Image
import scipy.special, tqdm
import numpy as np
import torchvision.transforms as transforms
from data.dataset import LaneTestDataset


# 改变输入视频分辨率
def change_video():
    flag_li = []
    video_li = [i for i in os.listdir(video_path) if i.endswith('.mp4') or i.endswith('.avi')]

    if len(video_li) == 0:
        return None

    for s in video_li:
        print(os.path.join(video_path, s))
        see_cap = cv2.VideoCapture(os.path.join(video_path, s))
        print(see_cap.get(3), see_cap.get(4))
        if int(see_cap.get(3)) == video_w and int(see_cap.get(4)) == video_h:
            flag_li.append(True)
        else:
            flag_li.append(False)

    if all(flag_li):
        return 0

    for li in video_li:
        print(os.path.join(video_path, li))
        video_cap = cv2.VideoCapture(os.path.join(video_path, li))

        video_fourcc = cv2.VideoWriter_fourcc(*'MJPG')
        video_save_path = os.path.join(temp_dir, li.split('.')[0] + '_' + str(video_w) + '_' + str(video_h) + '.avi')
        video_writer = cv2.VideoWriter(video_save_path, video_fourcc, 30.0, (video_w, video_h))
        success, _ = video_cap.read()
        while success:
            success, img1 = video_cap.read()
            try:
                img = cv2.resize(img1, (video_w, video_h), interpolation=cv2.INTER_LINEAR)
                video_writer.write(img)
            except:
                break
    return 1


if __name__ == "__main__":
    torch.backends.cudnn.benchmark = True
    args, cfg = merge_config()
    dist_print('start testing...')
    assert cfg.backbone in ['18', '34', '50', '101', '152', '50next', '101next', '50wide', '101wide']

    if cfg.dataset == 'CULane':
        cls_num_per_lane = 18
        video_w, video_h = 1640, 590
    elif cfg.dataset == 'Tusimple':
        cls_num_per_lane = 56
        video_w, video_h = 1280, 720
    else:
        raise NotImplementedError

    net = parsingNet(pretrained=False, backbone=cfg.backbone, cls_dim=(cfg.griding_num + 1, cls_num_per_lane, 4),
                     use_aux=False).cuda()  # we dont need auxiliary segmentation in testing

    state_dict = torch.load(cfg.test_model, map_location='cpu')['model']
    compatible_state_dict = {}
    for k, v in state_dict.items():
        if 'module.' in k:
            compatible_state_dict[k[7:]] = v
        else:
            compatible_state_dict[k] = v

    net.load_state_dict(compatible_state_dict, strict=False)
    net.eval()

    img_transforms = transforms.Compose([
        # transforms.Resize((288, 800)),
        transforms.Resize((video_h, video_w)),
        transforms.ToTensor(),
        transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
    ])

    video_path = cfg.video_path if cfg.video_path != None else os.path.abspath('.')
    temp_dir = os.path.join(video_path, 'after_handel')

    if not os.path.exists(temp_dir):
        os.makedirs(temp_dir)

    # 0: 表示原视频未处理
    # 1: 调试原视频被处理
    num = change_video()
    handel_path = video_path if num == 0 else temp_dir
    after_video_li = [i for i in os.listdir(handel_path) if i.endswith('.avi') or i.endswith('.mp4')]

    for v in after_video_li:
        full_video_path = os.path.join(handel_path, v)
        cap = cv2.VideoCapture(full_video_path)
        fourcc = cv2.VideoWriter_fourcc(*'MJPG')
        end_save_path = os.path.join(handel_path, v.split('.')[0] + '_ok' + '.avi')
        vout = cv2.VideoWriter(end_save_path, fourcc, 30.0, (video_w, video_h))
        print('加载cuda是否成功：',torch.cuda.is_available())
        while 1:
            rval,frame = cap.read()
            if rval == False:
                break

            img = cv2.cvtColor(frame,cv2.COLOR_BGR2RGB)
            img_ = Image.fromarray(img)
            imgs = img_transforms(img_)
            imgs = imgs.unsqueeze(0)
            imgs = imgs.cuda()
            with torch.no_grad():
                out = net(imgs)

            col_sample = np.linspace(0, 800 - 1, cfg.griding_num)
            col_sample_w = col_sample[1] - col_sample[0]

            out_j = out[0].data.cpu().numpy()
            out_j = out_j[:, ::-1, :]
            prob = scipy.special.softmax(out_j[:-1, :, :], axis=0)
            idx = np.arange(cfg.griding_num) + 1
            idx = idx.reshape(-1, 1, 1)
            loc = np.sum(prob * idx, axis=0)
            out_j = np.argmax(out_j, axis=0)
            loc[out_j == cfg.griding_num] = 0
            out_j = loc

            for i in range(out_j.shape[1]):
                if np.sum(out_j[:, i] != 0) > 2:
                    for k in range(out_j.shape[0]):
                        if out_j[k, i] > 0:
                            ppp = (int(out_j[k, i] * col_sample_w * video_w / 800) - 1, int(video_h - k * 20) - 1)
                            cv2.circle(frame, ppp, 5, (0, 255, 0), -1)
            vout.write(frame)

        vout.release()

**Comment by zchrissirhcz** [comment]
Hi, @SubePython 
Please note it is markdown rendered in the "write " area. It's a better practice to put your code into code region, i.e. begin with  3 "```" in first line, then paste your code, then end with a line starting with 3 "```".

**Comment by cfzd** [comment]
@SubePython 
1. ```transforms.Resize((video_h, video_w))``` should be ```transforms.Resize((288, 800))```. 
2. Use our updated ```demo.py```, it supports both CULane and Tusimple resolution.
3. The order of resizing should be: resize your video to CULane or Tusimple resolution, then use the `demo.py` corresponding to the dataset to get the output. Finally, resize the output t to the resolution of the video.

I also think that if you use ```transforms.Resize((video_h, video_w))```, the network should report an error. How did you solve this problem?

**Comment by SubePython** [comment]
@cfzd 
>*‘’I also think that if you use transforms.Resize((video_h, video_w)), the network should report an error. How did you solve this problem?“

**抱歉，并未解决这个问题，是我最后一次调试的时候，尝试更改的，您说不错，的确会报错，对您造成了困扰，不好意思_**
**Sorry, it didn't solve this problem. I tried to change it the last time I tried to debug it. You said it was good. It did report an error, which caused trouble to you. I'm sorry**

>1. transforms.Resize((video_h, video_w)) should be transforms.Resize((288, 800)).
>2. Use our updated demo.py, it supports both CULane and Tusimple resolution.
>3. The order of resizing should be: resize your video to CULane or Tusimple resolution, then use the demo.py corresponding to >the dataset to get the output. Finally, resize the output t to the resolution of the video.

**我的流程是先处理视频，将视频大小调整为CULane或Tusimple数据集对于的分表率，然后修改demo使其读取视频并处理，最后输出视频，还原原视频大小，但效果和上图一样的，我确定clone的代码是最新的**
**My process is to first process the video, adjust the video size to the scale ratio of the culane or tusimple dataset, then modify the demo to read and process the video, and finally output the video to restore the original video size. However, the effect is the same as the above figure. I confirm that the code of clone is up-to-date**

@zchrissirhcz 
>Please note it is markdown rendered in the "write " area. It's a better practice to put your code into code region, i.e. begin with 3 "" in first line, then paste your code, then end with a line starting with 3 "".

**感谢您的解答：
您讲的是上图中change_video函数吗，如果是的话，上面的函数我注释掉在main里不调用，最后输出的视频效果和上图基本一致的，它的作用仅仅是处理原视频分辨率的大小
如果您指的不是change_video函数，是什么呢？**
**Thank you for your answers
You are talking about change in the figure above_ Video function? If it is, the above function will not be called in main. The final output video effect is basically the same as the above figure. Its function is only to process the resolution of the original video
If you don't mean change_ What is the video function? * *

**Comment by cfzd** [comment]
@SubePython 
I'm wondering which model did you use? Are using incompatible model and config, i.e. Culane model and Tusimple config, when visualizing the results?

**Comment by vabrishami** [comment]
@SubePython Could you solve this problem? I have exactly the same issue and the suggestions did not solve it. Thanks for your help.
